rules/opentitan/cc.bzl

```diff
@@ -0,0 +1,329 @@
+        disabled_features: List of disabled/unsupported features.
+        name: Name of the output binary.
+        srcs: Sources to compile this binary.
+        copts: Compiler options.
+        defines: Define values to pass to the compiler.
+        local_defines: Define values to pass to the compiler.
+        includes: Include directories to pass to the compiler.
+        deps: Dependencies for this binary.
+        linker_script: Linker script for this binary.
+        linkopts: Linker options for this binary.
+    Returns:
+      (elf_file, map_file) File objects.
+    """
+    cc_toolchain = find_cc_toolchain(ctx).cc
+    features = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = get_override(ctx, "features", kwargs),
+        unsupported_features = get_override(ctx, "disabled_features", kwargs),
+    )
+
+    compilation_contexts = [
+        dep[CcInfo].compilation_context
+        for dep in get_override(ctx, "attr.deps", kwargs)
+    ]
+    linker_script = get_override(ctx, "attr.linker_script", kwargs)
+    if linker_script:
+        compilation_contexts.append(linker_script[CcInfo].compilation_context)
+
+    name = get_override(ctx, "attr.name", kwargs)
+    cctx, cout = cc_common.compile(
+        name = name,
+        actions = ctx.actions,
+        feature_configuration = features,
+        cc_toolchain = cc_toolchain,
+        compilation_contexts = compilation_contexts,
+        srcs = get_override(ctx, "files.srcs", kwargs),
+        # FIXME: make/location substitution
+        user_compile_flags = ["-ffreestanding"] + get_override(ctx, "attr.copts", kwargs),
+        defines = get_override(ctx, "attr.defines", kwargs),
+        local_defines = get_override(ctx, "attr.local_defines", kwargs),
+        quote_includes = get_override(ctx, "attr.includes", kwargs),
+    )
+
+    linking_contexts = [
+        dep[CcInfo].linking_context
+        for dep in get_override(ctx, "attr.deps", kwargs)
+    ]
+    if linker_script:
+        linking_contexts.append(linker_script[CcInfo].linking_context)
+    mapfile = kwargs.get("mapfile", "{}.map".format(name))
+    mapfile = ctx.actions.declare_file(mapfile)
+    linkopts = [
+        "-Wl,-Map={}".format(mapfile.path),
+        "-nostdlib",
+    ] + get_override(ctx, "attr.linkopts", kwargs)
+
+    lout = cc_common.link(
+        name = name,
+        actions = ctx.actions,
+        feature_configuration = features,
+        cc_toolchain = cc_toolchain,
+        compilation_outputs = cout,
+        linking_contexts = linking_contexts,
+        # FIXME: make/location substitution
+        user_link_flags = linkopts,
+        additional_outputs = [mapfile],
+    )
+
+    return lout.executable, mapfile
+
+def _as_group_info(name, items):
+    """Prepare a dict of files for OutputGroupInfo.
+
+    Args:
+      name: A prefix for each dictionary key.
+      items: A dict str:File to prepare.
+    Returns:
+      dict
```

Comment:
Can you document a bit more the names of the groups generated, or add an example?

Reply:
Done